packages/neo-one-monitor/src/BrowserMonitor.js

```diff
@@ -30,30 +55,193 @@ class BrowserGauge extends BaseMetric implements Gauge {
   observe(countOrLabels?: number | RawLabels, count?: number): void {}
 }
 
-class BrowserMetricsFactory implements MetricsFactory {
+interface MetricsFactoryCollect extends MetricsFactory {
+  collect(): Object;
+}
+
+type MetricCollection = {|
+  counters: Array<BaseMetric>,
+  gauges: Array<BaseMetric>,
+  histograms: Array<BaseMetric>,
+  summaries: Array<BaseMetric>,
+|};
+
+class BrowserMetricsFactory implements MetricsFactoryCollect {
+  _metrics: MetricCollection;
+
+  constructor() {
+    this._metrics = {
+      counters: [],
+      gauges: [],
+      histograms: [],
+      summaries: [],
+    };
+  }
+
+  collect(): MetricCollection {
+    const metrics = JSON.parse(JSON.stringify(this._metrics));
+    for (const metricType of Object.keys(this._metrics)) {
+      this._metrics[metricType].map(metric => metric.reset());
+    }
+
+    return metrics;
+  }
+
   createCounter(options: MetricConstruct): Counter {
-    return new BrowserCounter(options);
+    const counter = new BrowserCounter(options);
+    this._metrics.counters.push(counter);
+
+    return counter;
   }
 
   createGauge(options: MetricConstruct): Gauge {
-    return new BrowserGauge(options);
+    const gauge = new BrowserGauge(options);
+    this._metrics.gauges.push(gauge);
+
+    return gauge;
   }
 
   createHistogram(options: MetricConstruct): Histogram {
-    return new BrowserHistogram(options);
+    const histogram = new BrowserHistogram(options);
+    this._metrics.histograms.push(histogram);
+
+    return histogram;
   }
 
   createSummary(options: MetricConstruct): Summary {
-    return new BrowserSummary(options);
+    const summary = new BrowserSummary(options);
+    this._metrics.summaries.push(summary);
+
+    return summary;
+  }
+}
+
+interface LoggerCollect extends Logger {
+  collect(): Array<LoggerLogOptions>;
+}
+
+export class BrowserLogger implements LoggerCollect {
+  _logs: Array<LoggerLogOptions>;
+
+  constructor() {
+    this._logs = [];
+  }
+
+  log(options: LoggerLogOptions): void {
+    this._logs.push(options);
+  }
+
+  collect(): Array<LoggerLogOptions> {
+    const logs = this._logs;
+    this._logs = [];
+
+    return logs;
+  }
+
+  close(callback: () => void): void {
+    callback();
+  }
+}
+
+export class Reporter {
+  _logger: BrowserLogger;
+  _monitor: BrowserMonitor;
+  _timer: number;
+  _endpoint: string;
+  _backLogs: Array<LoggerLogOptions>;
+  _backMetrics: MetricCollection;
+  _shutdownFunc: () => void;
+
+  constructor({
+    logger,
+    monitor,
+    timer,
+    endpoint,
+  }: {|
+    logger: BrowserLogger,
+    monitor: BrowserMonitor,
+    timer: number,
+    endpoint: string,
+  |}) {
+    this._logger = logger;
+    this._monitor = monitor;
+    this._timer = timer;
+    this._endpoint = endpoint;
+    this._backLogs = [];
+    this._backMetrics = {
+      counters: [],
+      gauges: [],
+      histograms: [],
+      summaries: [],
+    };
+
+    const subscription = interval(this._timer).subscribe(async () => {
```

Comment:
use `pipe(concatMap(async () => {})).subscribe()` to ensure that the timers are processed serially. As it is now, it's possible that two invocations of your subscribe function happen simultaneously. In general, you almost never want to use `subscribe` directly.

Reply:
In fact, it might be a good exercise to try to translate this to use `.pipe(mergeScan(..., [], 1))` and eliminate the local state changes - only use the aggregated argument for `mergeScan`.